File: sdks/php/tests/ModelsTest.php
<?php

declare(strict_types=1);

namespace Renamed\Tests;

use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;
use Renamed\Models\ExtractResult;
use Renamed\Models\JobStatusResponse;
use Renamed\Models\PdfSplitResult;
use Renamed\Models\RenameResult;
use Renamed\Models\SplitDocument;
use Renamed\Models\Team;
use Renamed\Models\User;

final class ModelsTest extends TestCase
{
    // Team Model Tests

    #[Test]
    public function team_from_array_creates_instance_with_required_fields(): void
    {
        $data = [
            'id' => 'team_123',
            'name' => 'Acme Corp',
        ];

        $team = Team::fromArray($data);

        $this->assertSame('team_123', $team->id);
        $this->assertSame('Acme Corp', $team->name);
    }

    #[Test]
    public function team_to_array_returns_correct_structure(): void
    {
        $team = new Team('team_456', 'Test Team');

        $array = $team->toArray();

        $this->assertSame([
            'id' => 'team_456',
            'name' => 'Test Team',
        ], $array);
    }

    // User Model Tests

    #[Test]
    public function user_from_array_creates_instance_with_required_fields(): void
    {
        $data = [
            'id' => 'user_123',
            'email' => 'test@example.com',
        ];

        $user = User::fromArray($data);

        $this->assertSame('user_123', $user->id);
        $this->assertSame('test@example.com', $user->email);
        $this->assertNull($user->name);
        $this->assertNull($user->credits);
        $this->assertNull($user->team);
    }

    #[Test]
    public function user_from_array_creates_instance_with_all_fields(): void
    {
        $data = [
            'id' => 'user_456',
            'email' => 'john@example.com',
            'name' => 'John Doe',
            'credits' => 100,
            'team' => [
                'id' => 'team_789',
                'name' => 'Development',
            ],
        ];

        $user = User::fromArray($data);

        $this->assertSame('user_456', $user->id);
        $this->assertSame('john@example.com', $user->email);
        $this->assertSame('John Doe', $user->name);
        $this->assertSame(100, $user->credits);
        $this->assertNotNull($user->team);
        $this->assertSame('team_789', $user->team->id);
        $this->assertSame('Development', $user->team->name);
    }

    #[Test]
    public function user_from_array_casts_credits_to_int(): void
    {
        $data = [
            'id' => 'user_123',
            'email' => 'test@example.com',
            'credits' => '50', // String value
        ];

        $user = User::fromArray($data);

        $this->assertSame(50, $user->credits);
    }

    #[Test]
    public function user_to_array_returns_correct_structure(): void
    {
        $team = new Team('team_123', 'Test Team');
        $user = new User('user_456', 'test@example.com', 'Test User', 200, $team);

        $array = $user->toArray();

        $this->assertSame([
            'id' => 'user_456',
            'email' => 'test@example.com',
            'name' => 'Test User',
            'credits' => 200,
            'team' => [
                'id' => 'team_123',
                'name' => 'Test Team',
            ],
        ], $array);
    }

    #[Test]
    public function user_to_array_handles_null_team(): void
    {
        $user = new User('user_123', 'test@example.com');

        $array = $user->toArray();

        $this->assertNull($array['team']);
    }

    // RenameResult Model Tests

    #[Test]
    public function rename_result_from_array_creates_instance_with_required_fields(): void
    {
        $data = [
            'originalFilename' => 'document.pdf',
            'suggestedFilename' => '2025-01-15_Invoice_12345.pdf',
        ];

        $result = RenameResult::fromArray($data);

        $this->assertSame('document.pdf', $result->originalFilename);
        $this->assertSame('2025-01-15_Invoice_12345.pdf', $result->suggestedFilename);
        $this->assertNull($result->folderPath);
        $this->assertNull($result->confidence);
    }

    #[Test]
    public function rename_result_from_array_creates_instance_with_all_fields(): void
    {
        $data = [
            'originalFilename' => 'scan.pdf',
            'suggestedFilename' => '2025-01-15_AcmeCorp_Invoice.pdf',
            'folderPath' => 'Invoices/2025/January',
            'confidence' => 0.95,
        ];

        $result = RenameResult::fromArray($data);

        $this->assertSame('scan.pdf', $result->originalFilename);
        $this->assertSame('2025-01-15_AcmeCorp_Invoice.pdf', $result->suggestedFilename);
        $this->assertSame('Invoices/2025/January', $result->folderPath);
        $this->assertSame(0.95, $result->confidence);
    }

    #[Test]
    public function rename_result_from_array_casts_confidence_to_float(): void
    {
        $data = [
            'originalFilename' => 'doc.pdf',
            'suggestedFilename' => 'renamed.pdf',
            'confidence' => '0.85', // String value
        ];

        $result = RenameResult::fromArray($data);

        $this->assertSame(0.85, $result->confidence);
    }

    #[Test]
    public function rename_result_to_array_returns_correct_structure(): void
    {
        $result = new RenameResult(
            'original.pdf',
            'suggested.pdf',
            'Documents/Archive',
            0.9
        );

        $array = $result->toArray();

        $this->assertSame([
            'originalFilename' => 'original.pdf',
            'suggestedFilename' => 'suggested.pdf',
            'folderPath' => 'Documents/Archive',
            'confidence' => 0.9,
        ], $array);
    }

    // ExtractResult Model Tests

    #[Test]
    public function extract_result_from_array_creates_instance_with_data(): void
    {
        $data = [
            'data' => [
                'vendor' => 'Acme Corp',
                'amount' => 150.00,
                'date' => '2025-01-15',
            ],
            'confidence' => 0.92,
        ];

        $result = ExtractResult::fromArray($data);

        $this->assertSame([
            'vendor' => 'Acme Corp',
            'amount' => 150.00,
            'date' => '2025-01-15',
        ], $result->data);
        $this->assertSame(0.92, $result->confidence);
    }

    #[Test]
    public function extract_result_from_array_defaults_to_empty_data(): void
    {
        $data = [
            'confidence' => 0.5,
        ];

        $result = ExtractResult::fromArray($data);

        $this->assertSame([], $result->data);
    }

    #[Test]
    public function extract_result_from_array_defaults_confidence_to_zero(): void
    {
        $data = [
            'data' => ['key' => 'value'],
        ];

        $result = ExtractResult::fromArray($data);

        $this->assertSame(0.0, $result->confidence);
    }

    #[Test]
    public function extract_result_to_array_returns_correct_structure(): void
    {
        $result = new ExtractResult(['field' => 'value'], 0.88);

        $array = $result->toArray();

        $this->assertSame([
            'data' => ['field' => 'value'],
            'confidence' => 0.88,
        ], $array);
    }

    // SplitDocument Model Tests

    #[Test]
    public function split_document_from_array_creates_instance(): void
    {
        $data = [
            'index' => 0,
            'filename' => 'document_part1.pdf',
            'pages' => '1-5',
            'downloadUrl' => 'https://api.renamed.to/download/abc123',
            'size' => 1024000,
        ];

        $doc = SplitDocument::fromArray($data);

        $this->assertSame(0, $doc->index);
        $this->assertSame('document_part1.pdf', $doc->filename);
        $this->assertSame('1-5', $doc->pages);
        $this->assertSame('https://api.renamed.to/download/abc123', $doc->downloadUrl);
        $this->assertSame(1024000, $doc->size);
    }

    #[Test]
    public function split_document_from_array_casts_numeric_fields(): void
    {
        $data = [
            'index' => '2', // String
            'filename' => 'doc.pdf',
            'pages' => '10-15',
            'downloadUrl' => 'https://example.com/download',
            'size' => '512000', // String
        ];

        $doc = SplitDocument::fromArray($data);

        $this->assertSame(2, $doc->index);
        $this->assertSame(512000, $doc->size);
    }

    #[Test]
    public function split_document_to_array_returns_correct_structure(): void
    {
        $doc = new SplitDocument(
            1,
            'part2.pdf',
            '6-10',
            'https://api.renamed.to/dl/xyz',
            2048000
        );

        $array = $doc->toArray();

        $this->assertSame([
            'index' => 1,
            'filename' => 'part2.pdf',
            'pages' => '6-10',
            'downloadUrl' => 'https://api.renamed.to/dl/xyz',
            'size' => 2048000,
        ], $array);
    }

    // PdfSplitResult Model Tests

    #[Test]
    public function pdf_split_result_from_array_creates_instance(): void
    {
        $data = [
            'originalFilename' => 'combined.pdf',
            'totalPages' => 20,
            'documents' => [
                [
                    'index' => 0,
                    'filename' => 'part1.pdf',
                    'pages' => '1-10',
                    'downloadUrl' => 'https://api.renamed.to/dl/a',
                    'size' => 1000000,
                ],
                [
                    'index' => 1,
                    'filename' => 'part2.pdf',
                    'pages' => '11-20',
                    'downloadUrl' => 'https://api.renamed.to/dl/b',
                    'size' => 1100000,
                ],
            ],
        ];

        $result = PdfSplitResult::fromArray($data);

        $this->assertSame('combined.pdf', $result->originalFilename);
        $this->assertSame(20, $result->totalPages);
        $this->assertCount(2, $result->documents);
        $this->assertInstanceOf(SplitDocument::class, $result->documents[0]);
        $this->assertSame('part1.pdf', $result->documents[0]->filename);
        $this->assertSame('part2.pdf', $result->documents[1]->filename);
    }

    #[Test]
    public function pdf_split_result_from_array_handles_empty_documents(): void
    {
        $data = [
            'originalFilename' => 'empty.pdf',
            'totalPages' => 0,
        ];

        $result = PdfSplitResult::fromArray($data);

        $this->assertSame([], $result->documents);
    }

    #[Test]
    public function pdf_split_result_to_array_returns_correct_structure(): void
    {
        $doc = new SplitDocument(0, 'doc.pdf', '1-5', 'https://dl.com/x', 500000);
        $result = new PdfSplitResult('original.pdf', [$doc], 5);

        $array = $result->toArray();

        $this->assertSame('original.pdf', $array['originalFilename']);
        $this->assertSame(5, $array['totalPages']);
        $this->assertCount(1, $array['documents']);
        $this->assertSame('doc.pdf', $array['documents'][0]['filename']);
    }

    // JobStatusResponse Model Tests

    #[Test]
    public function job_status_response_from_array_creates_instance_with_required_fields(): void
    {
        $data = [
            'jobId' => 'job_abc123',
            'status' => 'pending',
        ];

        $response = JobStatusResponse::fromArray($data);

        $this->assertSame('job_abc123', $response->jobId);
        $this->assertSame('pending', $response->status);
        $this->assertNull($response->progress);
        $this->assertNull($response->error);
        $this->assertNull($response->result);
    }

    #[Test]
    public function job_status_response_from_array_creates_instance_with_progress(): void
    {
        $data = [
            'jobId' => 'job_xyz',
            'status' => 'processing',
            'progress' => 50,
        ];

        $response = JobStatusResponse::fromArray($data);

        $this->assertSame(50, $response->progress);
    }

    #[Test]
    public function job_status_response_from_array_creates_instance_with_error(): void
    {
        $data = [
            'jobId' => 'job_failed',
            'status' => 'failed',
            'error' => 'Processing failed due to corrupt PDF',
        ];

        $response = JobStatusResponse::fromArray($data);

        $this->assertSame('failed', $response->status);
        $this->assertSame('Processing failed due to corrupt PDF', $response->error);
    }

    #[Test]
    public function job_status_response_from_array_creates_instance_with_result(): void
    {
        $data = [
            'jobId' => 'job_complete',
            'status' => 'completed',
            'progress' => 100,
            'result' => [
                'originalFilename' => 'multi.pdf',
                'totalPages' => 10,
                'documents' => [
                    [
                        'index' => 0,
                        'filename' => 'split.pdf',
                        'pages' => '1-10',
                        'downloadUrl' => 'https://dl.com/file',
                        'size' => 100000,
                    ],
                ],
            ],
        ];

        $response = JobStatusResponse::fromArray($data);

        $this->assertNotNull($response->result);
        $this->assertInstanceOf(PdfSplitResult::class, $response->result);
        $this->assertSame('multi.pdf', $response->result->originalFilename);
    }

    #[Test]
    public function job_status_response_is_completed_returns_true_for_completed(): void
    {
        $response = new JobStatusResponse('job_1', 'completed');

        $this->assertTrue($response->isCompleted());
        $this->assertFalse($response->isFailed());
        $this->assertFalse($response->isProcessing());
    }

    #[Test]
    public function job_status_response_is_failed_returns_true_for_failed(): void
    {
        $response = new JobStatusResponse('job_2', 'failed', null, 'Error message');

        $this->assertFalse($response->isCompleted());
        $this->assertTrue($response->isFailed());
        $this->assertFalse($response->isProcessing());
    }

    #[Test]
    public function job_status_response_is_processing_returns_true_for_pending(): void
    {
        $response = new JobStatusResponse('job_3', 'pending');

        $this->assertFalse($response->isCompleted());
        $this->assertFalse($response->isFailed());
        $this->assertTrue($response->isProcessing());
    }

    #[Test]
    public function job_status_response_is_processing_returns_true_for_processing(): void
    {
        $response = new JobStatusResponse('job_4', 'processing', 50);

        $this->assertFalse($response->isCompleted());
        $this->assertFalse($response->isFailed());
        $this->assertTrue($response->isProcessing());
    }

    #[Test]
    public function job_status_response_to_array_returns_correct_structure(): void
    {
        $doc = new SplitDocument(0, 'output.pdf', '1-5', 'https://dl.com/f', 50000);
        $splitResult = new PdfSplitResult('input.pdf', [$doc], 5);
        $response = new JobStatusResponse('job_5', 'completed', 100, null, $splitResult);

        $array = $response->toArray();

        $this->assertSame('job_5', $array['jobId']);
        $this->assertSame('completed', $array['status']);
        $this->assertSame(100, $array['progress']);
        $this->assertNull($array['error']);
        $this->assertIsArray($array['result']);
        $this->assertSame('input.pdf', $array['result']['originalFilename']);
    }

    #[Test]
    public function job_status_response_to_array_handles_null_result(): void
    {
        $response = new JobStatusResponse('job_6', 'pending');

        $array = $response->toArray();

        $this->assertNull($array['result']);
    }
}

File: sdks/ruby/test/models_test.rb
# frozen_string_literal: true

require "test_helper"

class ModelsTest < Minitest::Test
  # RenameResult tests

  def test_rename_result_from_response
    data = {
      "originalFilename" => "document.pdf",
      "suggestedFilename" => "2025-01-15_Invoice.pdf",
      "folderPath" => "Invoices/2025",
      "confidence" => 0.95
    }

    result = Renamed::RenameResult.from_response(data)

    assert_equal "document.pdf", result.original_filename
    assert_equal "2025-01-15_Invoice.pdf", result.suggested_filename
    assert_equal "Invoices/2025", result.folder_path
    assert_equal 0.95, result.confidence
  end

  def test_rename_result_from_response_with_nil_optional_fields
    data = {
      "originalFilename" => "document.pdf",
      "suggestedFilename" => "renamed.pdf"
    }

    result = Renamed::RenameResult.from_response(data)

    assert_equal "document.pdf", result.original_filename
    assert_equal "renamed.pdf", result.suggested_filename
    assert_nil result.folder_path
    assert_nil result.confidence
  end

  def test_rename_result_to_h
    result = Renamed::RenameResult.new(
      original_filename: "test.pdf",
      suggested_filename: "renamed.pdf",
      folder_path: "Documents",
      confidence: 0.9
    )

    hash = result.to_h

    assert_equal "test.pdf", hash[:original_filename]
    assert_equal "renamed.pdf", hash[:suggested_filename]
    assert_equal "Documents", hash[:folder_path]
    assert_equal 0.9, hash[:confidence]
  end

  def test_rename_result_to_h_omits_nil_values
    result = Renamed::RenameResult.new(
      original_filename: "test.pdf",
      suggested_filename: "renamed.pdf"
    )

    hash = result.to_h

    assert_equal({ original_filename: "test.pdf", suggested_filename: "renamed.pdf" }, hash)
    refute hash.key?(:folder_path)
    refute hash.key?(:confidence)
  end

  # SplitDocument tests

  def test_split_document_from_response
    data = {
      "index" => 0,
      "filename" => "document_part1.pdf",
      "pages" => "1-5",
      "downloadUrl" => "https://example.com/download/123",
      "size" => 12345
    }

    doc = Renamed::SplitDocument.from_response(data)

    assert_equal 0, doc.index
    assert_equal "document_part1.pdf", doc.filename
    assert_equal "1-5", doc.pages
    assert_equal "https://example.com/download/123", doc.download_url
    assert_equal 12_345, doc.size
  end

  def test_split_document_to_h
    doc = Renamed::SplitDocument.new(
      index: 1,
      filename: "part2.pdf",
      pages: "6-10",
      download_url: "https://example.com/file",
      size: 5000
    )

    hash = doc.to_h

    assert_equal 1, hash[:index]
    assert_equal "part2.pdf", hash[:filename]
    assert_equal "6-10", hash[:pages]
    assert_equal "https://example.com/file", hash[:download_url]
    assert_equal 5000, hash[:size]
  end

  # PdfSplitResult tests

  def test_pdf_split_result_from_response
    data = {
      "originalFilename" => "multi-page.pdf",
      "totalPages" => 10,
      "documents" => [
        {
          "index" => 0,
          "filename" => "part1.pdf",
          "pages" => "1-5",
          "downloadUrl" => "https://example.com/1",
          "size" => 1000
        },
        {
          "index" => 1,
          "filename" => "part2.pdf",
          "pages" => "6-10",
          "downloadUrl" => "https://example.com/2",
          "size" => 2000
        }
      ]
    }

    result = Renamed::PdfSplitResult.from_response(data)

    assert_equal "multi-page.pdf", result.original_filename
    assert_equal 10, result.total_pages
    assert_equal 2, result.documents.length
    assert_instance_of Renamed::SplitDocument, result.documents[0]
    assert_equal "part1.pdf", result.documents[0].filename
    assert_equal "part2.pdf", result.documents[1].filename
  end

  def test_pdf_split_result_from_response_with_empty_documents
    data = {
      "originalFilename" => "empty.pdf",
      "totalPages" => 0
    }

    result = Renamed::PdfSplitResult.from_response(data)

    assert_equal "empty.pdf", result.original_filename
    assert_equal 0, result.total_pages
    assert_empty result.documents
  end

  def test_pdf_split_result_to_h
    result = Renamed::PdfSplitResult.new(
      original_filename: "test.pdf",
      total_pages: 5,
      documents: [
        Renamed::SplitDocument.new(
          index: 0,
          filename: "part1.pdf",
          pages: "1-5",
          download_url: "https://example.com/1",
          size: 1000
        )
      ]
    )

    hash = result.to_h

    assert_equal "test.pdf", hash[:original_filename]
    assert_equal 5, hash[:total_pages]
    assert_equal 1, hash[:documents].length
    assert_equal "part1.pdf", hash[:documents][0][:filename]
  end

  # JobStatusResponse tests

  def test_job_status_response_from_response_pending
    data = {
      "jobId" => "job_123",
      "status" => "pending",
      "progress" => 0
    }

    response = Renamed::JobStatusResponse.from_response(data)

    assert_equal "job_123", response.job_id
    assert_equal "pending", response.status
    assert_equal 0, response.progress
    assert_nil response.error
    assert_nil response.result
  end

  def test_job_status_response_from_response_processing
    data = {
      "jobId" => "job_123",
      "status" => "processing",
      "progress" => 50
    }

    response = Renamed::JobStatusResponse.from_response(data)

    assert_equal "job_123", response.job_id
    assert_equal "processing", response.status
    assert_equal 50, response.progress
  end

  def test_job_status_response_from_response_completed
    data = {
      "jobId" => "job_123",
      "status" => "completed",
      "progress" => 100,
      "result" => {
        "originalFilename" => "doc.pdf",
        "totalPages" => 5,
        "documents" => []
      }
    }

    response = Renamed::JobStatusResponse.from_response(data)

    assert_equal "job_123", response.job_id
    assert_equal "completed", response.status
    assert_equal 100, response.progress
    assert_instance_of Renamed::PdfSplitResult, response.result
  end

  def test_job_status_response_from_response_failed
    data = {
      "jobId" => "job_123",
      "status" => "failed",
      "error" => "Processing failed: invalid PDF"
    }

    response = Renamed::JobStatusResponse.from_response(data)

    assert_equal "job_123", response.job_id
    assert_equal "failed", response.status
    assert_equal "Processing failed: invalid PDF", response.error
  end

  def test_job_status_response_to_h
    response = Renamed::JobStatusResponse.new(
      job_id: "job_456",
      status: "processing",
      progress: 75
    )

    hash = response.to_h

    assert_equal "job_456", hash[:job_id]
    assert_equal "processing", hash[:status]
    assert_equal 75, hash[:progress]
  end

  # ExtractResult tests

  def test_extract_result_from_response
    data = {
      "data" => {
        "vendor" => "Acme Corp",
        "amount" => 150.00,
        "date" => "2025-01-15"
      },
      "confidence" => 0.88
    }

    result = Renamed::ExtractResult.from_response(data)

    assert_equal({ "vendor" => "Acme Corp", "amount" => 150.00, "date" => "2025-01-15" }, result.data)
    assert_equal 0.88, result.confidence
  end

  def test_extract_result_to_h
    result = Renamed::ExtractResult.new(
      data: { "key" => "value" },
      confidence: 0.95
    )

    hash = result.to_h

    assert_equal({ "key" => "value" }, hash[:data])
    assert_equal 0.95, hash[:confidence]
  end

  # Team tests

  def test_team_from_response
    data = {
      "id" => "team_789",
      "name" => "Engineering Team"
    }

    team = Renamed::Team.from_response(data)

    assert_equal "team_789", team.id
    assert_equal "Engineering Team", team.name
  end

  def test_team_to_h
    team = Renamed::Team.new(id: "team_123", name: "Test Team")

    hash = team.to_h

    assert_equal({ id: "team_123", name: "Test Team" }, hash)
  end

  # User tests

  def test_user_from_response
    data = {
      "id" => "user_123",
      "email" => "user@example.com",
      "name" => "Test User",
      "credits" => 500
    }

    user = Renamed::User.from_response(data)

    assert_equal "user_123", user.id
    assert_equal "user@example.com", user.email
    assert_equal "Test User", user.name
    assert_equal 500, user.credits
    assert_nil user.team
  end

  def test_user_from_response_with_team
    data = {
      "id" => "user_123",
      "email" => "user@example.com",
      "team" => {
        "id" => "team_456",
        "name" => "Acme Corp"
      }
    }

    user = Renamed::User.from_response(data)

    assert_instance_of Renamed::Team, user.team
    assert_equal "team_456", user.team.id
    assert_equal "Acme Corp", user.team.name
  end

  def test_user_to_h
    user = Renamed::User.new(
      id: "user_123",
      email: "test@example.com",
      name: "Test User",
      credits: 100
    )

    hash = user.to_h

    assert_equal "user_123", hash[:id]
    assert_equal "test@example.com", hash[:email]
    assert_equal "Test User", hash[:name]
    assert_equal 100, hash[:credits]
  end

  def test_user_to_h_with_team
    user = Renamed::User.new(
      id: "user_123",
      email: "test@example.com",
      team: Renamed::Team.new(id: "team_456", name: "My Team")
    )

    hash = user.to_h

    assert_equal "team_456", hash[:team][:id]
    assert_equal "My Team", hash[:team][:name]
  end

  def test_user_to_h_omits_nil_values
    user = Renamed::User.new(id: "user_123", email: "test@example.com")

    hash = user.to_h

    assert_equal({ id: "user_123", email: "test@example.com" }, hash)
    refute hash.key?(:name)
    refute hash.key?(:credits)
    refute hash.key?(:team)
  end

  # MIME_TYPES constant test

  def test_mime_types_constant
    assert_equal "application/pdf", Renamed::MIME_TYPES[".pdf"]
    assert_equal "image/jpeg", Renamed::MIME_TYPES[".jpg"]
    assert_equal "image/jpeg", Renamed::MIME_TYPES[".jpeg"]
    assert_equal "image/png", Renamed::MIME_TYPES[".png"]
    assert_equal "image/tiff", Renamed::MIME_TYPES[".tiff"]
    assert_equal "image/tiff", Renamed::MIME_TYPES[".tif"]
  end

  def test_mime_types_is_frozen
    assert Renamed::MIME_TYPES.frozen?
  end
end
